### DISCUSSION TITLE
Accountability for user created from ALLOW_PUBLIC_REGISTRATION flag

### DISCUSSION THREAD

**Discussion by gkielwasser** [directus/directus#11008] [Archive]
Hi,

when a user is created from `AUTH_*_ALLOW_PUBLIC_REGISTRATION` flag, there is no accountability.

Could we add it from the request it comes from ?

Here is where we don't pass it https://github.com/directus/directus/blob/c4d54ad34f45dad02c2a9d932355ebe2bf3c4877/api/src/auth/drivers/oauth2.ts#L40

**Comment by rijkvanzanten** [comment]
What do you mean there's no accountability? The system itself creates the user in this case, so there won't be an accountability object with things like user / role, as they don't exist 🤔

**Reply by gkielwasser** [reply]
Writting this long description my discussion title is not occurate anymore :)

Here is th exception I have and I try to understand. It happens when I try to create a Directus notification using the items service while using a built accountability object.
Just to clarify I'm using directly the items service so there is no logic about sending email that is used when I create a notification.

```
const context = { accountability }
const customNotificationsService = new ItemsService('directus_notifications', context)
await customNotificationsService.createOne({ ... })
```
 It seems I miss the `ip` in the accountability object. I don't get why I'm forced to give it. The exception thrown is:
```
Error: Value for field "ip" can't be null.
```

Here are the 2 differents registration flow where I'm using the upper notification. The method is called from the hooked in `action('users.create')` to notify a user has been created :

**1- register with username/password in a custom endpoint** 
In this method I create a user using the user service. I create an instance of the service using a custom accountablity object based on the request object and admin permissions. Writing this, it might be bad to mix admin data with request data related to the enduser. Maybe I should just use admin permissions in the accountability object.
```
const userService = new UsersService({
        schema: req.schema,
        accountability: {
            ...req.accountability,
            //  Give admin permissions
            admin: true,
            role: null,
            permissions: []
        }
    });
    await userService .createOne({ ... })
```
the accountability built based on the request looks like that
```
{
  ...
  "ip": "::1",
  "userAgent": "...",
  ...
}
```
In this case I have no exception because of the accountability I pass that is containing the ip field.

**2- register using Directus flow (oauth and public registration flag)**
In this flow no accountability exist because none has been given when creating the user. 
https://github.com/directus/directus/blob/c4d54ad34f45dad02c2a9d932355ebe2bf3c4877/api/src/auth/drivers/oauth2.ts#L40
I thought we could have request informations related to the user even if we don't have an authenticated user. Something like this:
```
{
  ...
  "user": null
  "ip": req.ip,
  "userAgent": req.useragent
  ...
}
```
So in this flow, I have the exception described upper that is thrown.

**Comment by gkielwasser** [comment]
For now I just ensure the accountability is null here to cover all my use cases
```
const customNotificationsService = new ItemsService('directus_notifications', {
  ...context,
  accountability: null
})
```
